File: js-lang/src/scope.ts
import { NarrowedSymbol, SourceSymbol, Symbol } from "./symbols";

export class Scope {
  public static empty = new Scope(null, []);

  constructor(
    private readonly parent: Scope | null = null,
    private readonly symbols: Symbol[] = []
  ) {}

  withSymbols(symbols: Symbol[]) {
    return new Scope(this.parent, symbols.concat(this.symbols));
  }

  resolveLast(name: string): Symbol | null {
    return (
      this.symbols.find((x) => x.name === name) ??
      this.parent?.resolveLast(name) ??
      null
    );
  }

  resolveAll(name: string): Symbol[] {
    return this.symbols
      .filter((x) => x.name === name)
      .concat(this.parent?.resolveAll(name) ?? []);
  }

  getAll(filter: "skip-duplicates" = "skip-duplicates"): Symbol[] {
    const allSymbols = this.symbols.concat(this.parent?.getAll() ?? []);

    // Filter out original symbols that has been narrowed
    for (let i = allSymbols.length - 1; i >= 0; i--) {
      const symbol = allSymbols[i];
      for (let j = 0; j < i; j++) {
        const laterSymbol = allSymbols[j];
        if (
          symbol instanceof SourceSymbol &&
          laterSymbol instanceof NarrowedSymbol &&
          symbol === laterSymbol.outerSymbol
        ) {
          allSymbols.splice(i, 1);
          break;
        }
      }
    }

    return allSymbols;
  }
}

File: dotnet/Tybscri/Common/Scope.cs
using System.Collections.Immutable;
using Tybscri.Symbols;

namespace Tybscri.Common;

public record Scope
{
    public static readonly Scope Empty = new Scope();

    private readonly Scope? _parent = null;
    private readonly ImmutableList<ISymbol> _symbols = ImmutableList<ISymbol>.Empty;

    public Scope()
    {
    }

    public Scope(IEnumerable<ISymbol> symbols) : this(null, symbols)
    {
    }

    private Scope(Scope? parent, IEnumerable<ISymbol> symbols)
    {
        _parent = parent;
        _symbols = symbols.ToImmutableList();
    }

    public ISymbol? ResolveLast(string name)
    {
        return _symbols.Find(x => x.Name == name) ?? _parent?.ResolveLast(name);
    }

    public Scope CreateChildScope(IEnumerable<ISymbol> symbols)
    {
        return new Scope(this, symbols);
    }

    public Scope WithSymbol(SourceSymbol sourceSymbol)
    {
        return new Scope(_parent, _symbols.Add(sourceSymbol));
    }
}